registry/manifest.go

```diff
@@ -0,0 +1,76 @@
+package registry
+
+import (
+	"encoding/json"
+	"errors"
+	"golang.org/x/net/context"
+	"golang.org/x/net/context/ctxhttp"
+	"net/http"
+)
+
+// ErrBadSchema is returned when server returns an unsupported or invalid schema format.
+var ErrBadSchema = errors.New("Server did not return a supported schema format")
+
+// ErrDigestNotFound is returned when server does not include digest header in response.
+var ErrDigestNotFound = errors.New("Docker-Content-Digest header not found")
+
+type manifestResponse struct {
+	SchemaVersion int    `json:"schemaVersion"`
+	MediaType     string `json:"mediaType"` // v2 only
+}
+
+const (
+	schemaV1MimeType = "application/vnd.docker.distribution.manifest.v1+json"
+	schemaV2MimeType = "application/vnd.docker.distribution.manifest.v2+json"
+)
+
+func (registry *Registry) GetManifestDigest(ctx context.Context, repository string, reference string) (string, error) {
+	url := registry.url("/v2/%s/manifests/%s", repository, reference)
+	registry.Logf("registry.manifests url=%s repository=%s reference=%s", url, repository, reference)
+
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return "", err
+	}
+
+	// both schema formats are valid
+	req.Header["Accept"] = []string{schemaV2MimeType, schemaV1MimeType}
+
+	resp, err := ctxhttp.Do(ctx, registry.Client, req)
+	if resp != nil {
+		defer resp.Body.Close()
+	}
+	if err != nil {
+		return "", chooseError(ctx, err)
+	}
+
+	var manifest manifestResponse
+	decoder := json.NewDecoder(resp.Body)
+	if err := decoder.Decode(&manifest); err != nil {
+		return "", err
+	}
+
+	// sanity check to at least make sure we are getting an image manifest
```

Comment:
My understanding is that we get SchemaVersion 1 if the manifest uploaded in version 1. Registry does convert version 2 to version 1 if we don't supply the `Accept` header, however there's no reverse conversion. 

Here's the registry code that serves GetManifest. Looks like there's only downgrade conversion (line 128 and onward). 
https://github.com/docker/distribution/blob/8907f7d/registry/handlers/images.go

In the same file, `PutManifest` unmarshalls a manifest (mostly just JSON decode), and doesn't seem to perform any conversion before writing. 

In my experience, `_/ubuntu` returns V1, whereas my own repo `ccp0101/appbase-node-0.10.45` returns V2.